File: src2/LU.py
EPS = 1e-12

# A.x = b
# A = L.U
# Ux = y
# Ly = b


def verificador(A, n):
    for i in range(n):
        if all(abs(A[i][j]) < EPS for j in range(n)):
            return False
    return True


def pivoteamento(k, n, b, A, L):
    maior = A[k][k]
    indice = k

    for i in range(k + 1, n):
        if abs(maior) < abs(A[i][k]):
            maior = A[i][k]
            indice = i

    # matriz singular detectada já no pivoteamento
    if abs(A[indice][k]) < EPS:
        return False

    if indice != k:
        # troca linhas em A
        for j in range(n):
            A[k][j], A[indice][j] = A[indice][j], A[k][j]

        # troca b
        b[k], b[indice] = b[indice], b[k]

        # troca as colunas já preenchidas de L (0 até k-1)
        for j in range(k):
            L[k][j], L[indice][j] = L[indice][j], L[k][j]

    return True


def gauss(n, k, b, A, L):
    for i in range(k + 1, n):
        coef = A[i][k] / A[k][k]
        for j in range(k, n):
            A[i][j] = A[i][j] - coef * A[k][j]

        # no LU o b não muda aqui
        L[i][k] = coef


def matriz_y(n, L, b, y):
    # L.y = b
    for i in range(n):
        soma = 0
        for j in range(i):
            soma += L[i][j] * y[j]
        y[i] = (b[i] - soma) / L[i][i]


def matriz_x(n, A, y, x):
    # U.x = y
    for i in range(n - 1, -1, -1):
        soma = 0
        for j in range(i + 1, n):
            soma += A[i][j] * x[j]
        x[i] = (y[i] - soma) / A[i][i]


def main():
    try:
        n = int(input("digite o numero de variáveis: "))
    except ValueError:
        print("Entrada inválida.")
        return

    if n <= 0:
        print("Número de variáveis inválido.")
        return

    A = [[0.0] * n for _ in range(n)]
    L = [[0.0] * n for _ in range(n)]

    b = [0.0] * n
    y = [0.0] * n
    x = [0.0] * n

    for i in range(n):
        for j in range(n):
            try:
                coef_a = float(input(f"digite o valor do coef a{i}{j}: "))
            except ValueError:
                print("Entrada inválida.")
                return
            A[i][j] = coef_a

    for i in range(n):
        try:
            B = float(input(f"digite o valor do coef b{i}: "))
        except ValueError:
            print("Entrada inválida.")
            return
        b[i] = B

    # verifica linhas nulas
    if not verificador(A, n):
        print("Linha nula detectada. Sistema inválido.")
        return

    # monta diagonal da matriz L
    for i in range(n):
        L[i][i] = 1

    # Fatoração LU com pivoteamento parcial
    for k in range(n - 1):
        if not pivoteamento(k, n, b, A, L):
            print("Matriz singular. Sistema não tem solução única.")
            return
        if abs(A[k][k]) < EPS:
            print("Matriz singular. Sistema não tem solução única.")
            return
        gauss(n, k, b, A, L)

    # verifica último pivô
    if abs(A[n-1][n-1]) < EPS:
        print("Matriz singular. Sistema não tem solução única.")
        return

    # resolve Ly = b
    matriz_y(n, L, b, y)

    # resolve Ux = y
    matriz_x(n, A, y, x)

    print("\nSolucao:")
    for i in range(n):
        print(f"x{i + 1} = {x[i]}")


if __name__ == "__main__":
    main()

File: src2/LU.cpp
#include <iostream>
#include <cmath>
#include <vector>
using namespace std;

//A.x=b
//A= L.U
//Ux= y 
//Ly = b

//pega matriz a -> transforma em matriz triangular inferior (L) e guarda os coeficientes


void verificador (vector<vector<double>> &A, int n) {
    for (int i = 0; i < n; i++){
        int zeros = 0;
        for (int j  = 0; j < n; j++){
            if (A[i][j] != 0){
                break;
            }
            zeros++;
        }
        if(zeros == n){
            cout << "linha nula, inválido";
            return;
        }
    }
}




void pivoteamento (int k, int n, vector<double>&b, vector<vector<double>>&A){
//partindo da linha que eu to eu vejo se tem algum na mesma coluna em uma linha mai spra baixo com valor maior, 
//se tiver, troca a linha e troca no b tb 
double maior = A[k][k];
int indice = k;
for (int i = k; i < n; i++){
    if (abs(maior) < abs(A[i][k])){
        maior = A[i][k];
        indice = i;
        }
    }



if (maior != A[k][k]){
//será feito pivoteamento 
for (int j = k; j < n; j++){ //varrendo a linha 
    double aux = A[k][j];
    A[k][j] = A[indice][j];
    A[indice][j] = aux;
    }
    double auxb = b[k];
    b[k] = b[indice];
    b[indice] = auxb;

//código do coeficiente da matriz l (nao pensei aind)

    }
}



void gauss (int n, int k,  vector<double>&b, vector<vector<double>>&A, vector<vector<double>>&L){
    //apos o pivoteamento, preciso zerar todos da coluna k que estou. 
    //linha_seguinte = linha_seguinte - coef x linha_atual
    //coef = linha_seguinte/linha_atual

    for (int i = k+1; i < n; i++){
        double coef = A[i][k]/A[k][k]; 
        for (int j = k; j < n; j++){
            A[i][j] = A[i][j] - coef * A[k][j];
        }

        //b[i] = b[i] - coef * b[k];
        L[i][k] = coef;
        //no lu o b nao muda aqui
    }
}



void matriz_y(int n, vector<vector<double>>&L,vector<double>&b, vector<double>&y){
//L.y = b //b[i] = A[linha i]
    for (int i = 0; i < n; i++) {//matriz triangular inferior

        double soma = 0;

        for (int j = 0; j < i; j++) {
            soma += L[i][j] * y[j];
        }

        y[i] = (b[i] - soma) / L[i][i];
    }
}




void matriz_x(int n, vector<vector<double>>&A,vector<double>&y, vector<double>&x){
//U(no caso matriz a).x=y

    
    for (int i = n - 1; i >= 0; i--) {//matriz triangular superiro

        double soma = 0;

        for (int j = i + 1; j < n; j++) {
            soma += A[i][j] * x[j];
        }

        x[i] = (y[i] - soma) / A[i][i];
    }
}












int main (){
    int n;
    cout << "digite o numero de variáveis: ";
    cin >> n;

    //matriz A e L
    vector<vector<double>>A(n, vector<double>(n));
    vector<vector<double>>L(n, vector<double>(n));

    //vetor b 
    vector<double>b(n);
    vector<double>y(n);
    vector<double>x(n);
    
    for (int i = 0; i < n; i++ ){
        for (int j = 0; j < n; j++){
            cout << "digite o valor do coef a" << i << j<< ": ";
            double coef_a;
            cin >> coef_a;
            A[i][j]=coef_a;
        }
    }

      for (int i = 0; i < n; i++ ){
        cout << "digite o valor do coef b" << i<< ": ";
        double B;
        cin >> B;
        b[i] = B;
      }

 // verifica linhas nulas
    verificador(A, n);

    //monta diagonal da matriz dos coef
    for (int i = 0; i < n; i++) {
        L[i][i] = 1;
    }

    // Fatoracao LU
    for (int k = 0; k < n - 1; k++) {

        pivoteamento(k, n, b, A);

        gauss(n, k, b, A, L);
    }

    // resolve Ly = b
    matriz_y(n, L, b, y);

    //resolve Ux = y
    matriz_x(n, A, y, x);

    cout << "\nSolucao:\n";

    for (int i = 0; i < n; i++) {
        cout << "x" << i + 1 << " = " << x[i] << endl;
    }

    return 0;
}




